### ISSUE TITLE
xonsh as a series of tubes

### ISSUE THREAD

**Issue by CJ-Wright** [xonsh/xonsh#932]
## Abstract

I would like to have a way to take a xonsh session and edit a version of the history to make a pipeline or a pipeline fragment. These fragments can be put together to make a complete data processing/analysis pipeline (although you don't need to just process data, you could use it for most anything). The provenance provided by xonsh's very rich and detailed history will make this possible, allowing for the translation between pipeline fragment name spaces. While the motivating goal of this is/was to generate a reproducible analysis pipelines, which could be used in a map/reduce context, this should be very applicable to other issues where one would like to pipeline a system that was initially designed/produced/tested by playing around in a xonsh shell like DevOps.
## Background

My background is in high energy x-ray scattering experiments. There is a lot of data processing and analysis. Additionally, not everyone uses the same software/scripts for data processing and analysis. Even my own scripts change over time, causing all sorts of provenance issues.
At odds with the concerns for provenance is the idea of data investigation & introspection. This is usually done in some sort of interactive  terminal (generally an IPython or Jupyter terminal). While these terminals work well for playing with the data, providing real time response to commands, easy graphing, and other joys associated with real time programming/execution, they do not easily produce scripts or script fragments, or capture the environment and execution context which are requirements for creation of a pipeline which can process all of your data (now that you have decided that you have the correct way to process it).
## Proposal

Thus, I propose a system to address these issues by:
1. Generate a pipeline/pipeline fragment from a xonsh session
   1. User plays with xonsh
   2. User then edits their work by editing a history like thing
   3. User specifies fragment inputs and outputs
2. The pipeline can then be put together
   1. Pipeline may work inside a xonsh session
   2. Pipeline is setup by putting fragment (names, hashes, uuids?) together with namespace translations in between
   3. Pipeline is checked for leaks; comparing the number of inputs for the n+1th fragment to the outputs of the nth fragment.
   4. If everything is ok (all the ins and outs match) run the pipeline
   5. Essentially each fragment will get a copy of the active xonsh environment and other history details, modified by the fragment itself (`a=1` becomes `a=2` after +1 in the nth fragment) and the translation layer (the nth fragment had an output named x but the n+1th fragment has that same variable named y)
   6. Notes :
      1. Trimming the namespace may be a good thing, if variables are not ever used, however we will need the ability to pass only parts of the xonsh environment to the pipeline, essentially allowing pieces of the environment which are not needed for a given fragment (but are needed later) to be oblivious to that fragments operation.
      2. Need to allow output kwargs to become input args if they have the same name and not enough args were given.
3. Collaboration, reproducibility, and beyond
   1. The pipelines all have uuids so that they can be used reproducibly on data and passed around.
   2. Need some sort of way to get fragments to other people
   3. Need some way to search for fragments to be used.
   4. It would be good to provide some sort of interface or something to map/reduce, dask, or pySpark since these pipelines should be able to run in parallel over large swaths of data or parameter space.
   5. It would also be nice to be able to run the pipeline as a stand alone script, especially for HPC applications.
## Conclusions

Producing reproducible pipelines from interactive programming sessions is an important task for all sorts of projects from data analysis to DevOps. Xonsh's unique history capacity makes it a prime candidate for this kind of work.

Thoughts?
In collaboration with @scopatz

**Comment by scopatz** [comment]
Thanks for putting this in @CJ-Wright!

**Comment by tacaswell** [comment]
Been thinking about this on the train and am almost to my stop so speed comments.

There are 3 parts to this (that I think can be handled semi-independently), pipeline capture/creation, pipeline manipulation and pipeline execution. 

There is already a lot of work in the space of representing and running pipelines (organge, vistrails, ITK (its own c++ y way), I know jupyter is interested, and my own poor entry pyrafters).  I would take very through looks at those before implementing that again.

I suspect xonsh can have the greatest impact on the capture/creation step.

**Comment by scopatz** [comment]
Hi @tacaswell - yeah, I completely agree with your three parts.  The notion of having external manipulation and execution came up yesterday in discussion as well. There are many graph execution frameworks and task schedulers out there and it would be great (ideal?) to interface with those - Spark, dask, etc.

The way this was proposed yesterday was for xonsh itself to have basic manipulation and execution utilities. This would be enough to run pipelines on laptops or clusters and would work for most people who don't want to have external dependencies or who want to do something relatively simple.

Then xonsh could also have hooks and support for these other pipeline execution tools, as many as folks are interested in supporting. These would certainly be welcome in the xonsh mainline, or they could be separate 3rd party packages.

I think it would be really neat to have.

**Comment by scopatz** [comment]
Also, and idea: filter out failed commands in the creation phase.

**Comment by tacaswell** [comment]
Filtering out 'failed' commands can be dangerous if they have side effects
(updating databases ect) and partially succeed

**Comment by scopatz** [comment]
Yeah, sorry for being curt here. I envisioned the filtering out failed commands as an optional activity.

**Comment by scopatz** [comment]
Closing due to age.  I still think this would be nice, but perhaps it could be a xontrib or another utility.
